mne/io/curry/curry.py

```diff
@@ -0,0 +1,317 @@
+from ..constants import FIFF
+from ...utils import check_fname, check_version, logger, verbose
+from ...annotations import Annotations
+
+INFO_FILE_EXTENSION = {7: '.dap', 8: '.cdt.dpa'}
+LABEL_FILE_EXTENSION = {7: '.rs3', 8: '.cdt.dpa'}
+DATA_FILE_EXTENSION = {7: '.dat', 8: '.cdt'}
+EVENT_FILE_EXTENSION = {7: '.cef', 8: '.cdt.cef'}
+CHANTYPES = {"meg": "_MAG1", "eeg": "", "misc": "_OTHERS"}
+FIFFV_CHANTYPES = {"meg": FIFF.FIFFV_MEG_CH, "eeg": FIFF.FIFFV_EEG_CH,
+                   "misc": FIFF.FIFFV_MISC_CH}
+SI_UNITS = dict(V=FIFF.FIFF_UNIT_V, T=FIFF.FIFF_UNIT_T)
+SI_UNIT_SCALE = dict(c=1e-2, m=1e-3, u=1e-6, μ=1e-6, n=1e-9, p=1e-12, f=1e-15)
+
+
+def _get_curry_version(file_extension):
+    """Check out the curry file version."""
+    return 8 if 'cdt' in file_extension else 7
+
+
+def _check_missing_files(fname_base, curry_vers):
+    """Check if all necessary files exist."""
+    _msg = "The following required files cannot be found: {0}.\nPlease make " \
+           "sure all required files are located in the same directory."
+
+    missing = [fname_base + ext
+               for ext in [DATA_FILE_EXTENSION[curry_vers],
+                           INFO_FILE_EXTENSION[curry_vers],
+                           LABEL_FILE_EXTENSION[curry_vers]]
+               if not op.isfile(fname_base + ext)]
+
+    if missing:
+        raise FileNotFoundError(_msg.format(np.unique(missing)))
+
+
+def _read_curry_lines(fname, regex_list):
+    """Read through the lines of a curry parameter files and save data.
+
+    Parameters
+    ----------
+    fname : str
+        Path to a curry file.
+    regex_list : list of str
+        A list of strings or regular expressions to search within the file.
+        Each element `regex` in `regex_list` must be formulated so that
+        `regex + " START_LIST"` initiates the start and `regex + " END_LIST"`
+        initiates the end of the elements that should be saved.
+
+    Returns
+    -------
+    data_dict : dict
+        A dictionary containing the extracted data. For each element `regex`
+        in `regex_list` a dictionary key `data_dict[regex]` is created, which
+        contains a list of the according data.
+
+    """
+    save_lines = {}
+    data_dict = {}
+
+    for regex in regex_list:
+        save_lines[regex] = False
+        data_dict[regex] = []
+
+    with open(fname) as fid:
+        for line in fid:
+            for regex in regex_list:
+                if re.match(regex + " END_LIST", line):
+                    save_lines[regex] = False
+
+                if save_lines[regex] and line != "\n":
+                    result = line.replace("\n", "")
+                    if "\t" in result:
+                        result = result.split("\t")
+                    data_dict[regex].append(result)
+
+                if re.match(regex + " START_LIST", line):
+                    save_lines[regex] = True
+
+    return data_dict
+
+
+def _read_curry_info(fname_base, curry_vers):
```

Comment:
Based on my recent comments and having in mind that we would like to
write `_read_annotations_curry` like this:

```py
def _read_annotations_curry(fname, sfreq='auto'):
    latency, description = _read_events_curry(fname)
    sfreq = _get_sfreq(fname) if sfreq == 'auto' else sfreq

    return Annotations(
        orig_time=None,
        onset=latency / sfreq,
        duration=np.zeros_like(latency),
        description=description,
    )
```

If we have `_get_curry_file_structure` that checks the files etc, we could rewrite
`_read_curry_info` that does what its name says (read curry info) and nothing else
we could write `_get_sfreq` like this:

```py
def _get_freq(fname):
    curry_file = _get_curry_file_structure(fname)
    return _read_curry_info(curry_file).sfreq 
```

Then you either construct the `info` in `__init__`  or  create a `_get_info`:

```py

ParsedInfo = namedtuple('ParsedInfo', 'sfreq, n_samples, is_ascii, foo, bar')
def _read_curry_info(fname):
    """Extract info from curry parameter files."""
    var_names = ['NumSamples', 'SampleFreqHz',
                 'DataFormat', 'SampleTimeUsec',
                 'NUM_SAMPLES', 'SAMPLE_FREQ_HZ',
                 'DATA_FORMAT', 'SAMPLE_TIME_USEC']

    param_dict = dict()
    unit_dict = dict()
    with open(fname) as fid:
        for line in iter(fid):
            if any(var_name in line for var_name in var_names):
                key, val = line.replace(" ", "").replace("\n", "").split("=")
                param_dict[key.lower().replace("_", "")] = val
            for type in CHANTYPES:
                if "DEVICE_PARAMETERS" + CHANTYPES[type] + " START" in line:
                    data_unit = next(fid)
                    unit_dict[type] = data_unit.replace(" ", "") \
                        .replace("\n", "").split("=")[-1]

    sfreq = float(param_dict["samplefreqhz"])
    time_step = float(param_dict["sampletimeusec"]) * 1e-6

    if (sfreq == 0) and (time_step != 0):
        sfreq = 1. / time_step

    return ParsedInfo(
        sfreq=sfreq,
        n_samples=float(param_dict['samplefreqhz']),
        ..
        is_ascii=(param_dict["dataformat"] == "ASCII")
    )


def _get_info(curry: CurryFileStructure) -> info_dict
    parsed_info = _read_curry_info(curry.info)
    labels = _read_curry_lines(curry.label,
                               ["LABELS" + CHANTYPES[key] for key in
                                ["meg", "eeg", "misc"]])

    sensors = _read_curry_lines(curry.label,
                                ["SENSORS" + CHANTYPES[key] for key in
                                 ["meg", "eeg", "misc"]])

    all_chans = list()
    for key in ["meg", "eeg", "misc"]:
        for ind, chan in enumerate(labels["LABELS" + CHANTYPES[key]]):
            ch = {"ch_name": chan,
                  "unit": unit_dict[key],
                  "kind": FIFFV_CHANTYPES[key]}
            if key in ("meg", "eeg"):
                loc = sensors["SENSORS" + CHANTYPES[key]][ind]
                ch["loc"] = np.array(loc, dtype=float)

            all_chans.append(ch)

    ch_names = [chan["ch_name"] for chan in all_chans]
    info = create_info(ch_names, parsed_info.sfreq)

    for ind, ch_dict in enumerate(info["chs"]):
        ch_dict["kind"] = all_chans[ind]["kind"]
        ch_dict['unit'] = SI_UNITS[all_chans[ind]['unit'][1]]
        ch_dict['cal'] = SI_UNIT_SCALE[all_chans[ind]['unit'][0]]
        if ch_dict["kind"] in (FIFF.FIFFV_MEG_CH,
                               FIFF.FIFFV_EEG_CH):
            ch_dict["loc"] = all_chans[ind]["loc"]

    return info, n_samples, parsed_info.is_ascii

```